### PR TITLE
Emit proper function prototypes

### PR DESCRIPTION
**Pull request by sergioferreiradt**
Last version emited public functions to interface definition without return value and proper parameter definitions.
This modification aims to emit the corresponding complete typescript method prototype.

### CODE REVIEW

File: Source/Modules/javascript.cxx [hunk 1]
```diff
@@ -2852,7 +2847,22 @@ int TypeScriptTypes::memberfunctionHandler(Node *n)
 {
   if (generateTsTypes)
   {
-    tsTypeInterfaceDeclaration->addMemberFunction(n);
+    String *functionName = Getattr(n, "sym:name");
+    String *functionPrototype = NewString(functionName);
+    Append(functionPrototype,"?(");
```
Comment (vadz) [inline Source/Modules/javascript.cxx]: It's a bit difficult to see what is this supposed to generate (a comment would probably help), but why should we always have `?` in the prototype?

File: Source/Modules/javascript.cxx [hunk 2]
```diff
@@ -2852,7 +2847,22 @@ int TypeScriptTypes::memberfunctionHandler(Node *n)
...
+    ParmList *l = Getattr(n, "parms");
+    Parm *p;
+    int i;
+    for (i = 0, p = l; p; i++) {
+      if ( i > 0) {
+        Append(functionPrototype,", ");
+      }
```
Comment (vadz) [inline Source/Modules/javascript.cxx]: This is a bit confusing as `i` is not really used as the loop index. I'd do it like this:
```suggestion
    ParmList* const l = Getattr(n, "parms");
    for (Perm* p = l; p; p = nextSibling(p)) {
      if ( p != l ) {
        Append(functionPrototype, ", ");
      }
```
(we could make `l` internal to the loop too, but this is probably not worth it -- but renaming it to `parms` might be).

File: Source/Modules/javascript.cxx [hunk 3]
```diff
@@ -2852,7 +2847,22 @@ int TypeScriptTypes::memberfunctionHandler(Node *n)
...
+      Printf(functionPrototype, "%s:%s", Getattr(p,"name"), getTypescriptType(p));
+      p = nextSibling(p);
+    }
+    String *typescriptType = getTypescriptType(n);
+    Printf(functionPrototype, ") : %s;", typescriptType);
```
Comment (vadz) [inline Source/Modules/javascript.cxx]: Very minor, but why bother with a temp variable?
```suggestion
    Printf(functionPrototype, ") : %s;", getTypescriptType(n));
```

File: Examples/test-suite/javascript/typescript_methods_runme.ts [hunk 4]
```diff
@@ -0,0 +1,10 @@
+import { C } from './typescript-methods-types';
+
+const typescript_interface = require("./typescript_methods");
+
+const c: C = new typescript_interface.C();
+c.intValue = 3;
+c.floatValue = 2.45;
+const s = c.sum(2,3);
```
Comment (vadz) [inline Examples/test-suite/javascript/typescript_methods_runme.ts:8]: Why not test the 2 other functions defined in `C`? As long as they're declared there, I think they should be tested.

File: Source/Modules/javascript.cxx (hunk 1 above)
Comment (sergioferreiradt) [reply Source/Modules/javascript.cxx]: If we don't mark the function as optional, then we cannot assign a variable directly with JSON. Probably this should be controlled by a pragma such as "typescriptoptionalmethods"

Comment (vadz) [reply Source/Modules/javascript.cxx]: I'm not sure I understand this. This should be generic code, not related to JSON in any way. Perhaps JSON structure needs to be checked, i.e. it should be explicitly asserted to contain the expected values in our code instead?

Comment (sergioferreiradt) [reply Source/Modules/javascript.cxx]: Sorry, probably I am complicating again :-(. 
Imagine we have following interface definition:

export interface C {
   intValue: number;
   floatValue: number;
   funcWithClass(cVar:C) : A;
}

Due to the function not being optional, this prevents us to do following initialization, Because of the non optional nature of funcWithClass():

const cValue: C = {
    intValue: 1,
    floatValue: 2.3,
};

I use initializations in unit tests. If the functions are not optional, then I cannot use the type when declaring the variable. Probably we wont use that much literal object initializations (sorry: I used the term JSON but the correct is literal object initialization)

Comment (vadz) [reply Source/Modules/javascript.cxx]: I'm not a TS expert but it seems strange to make all functions optional to me. And how/what with are they _supposed_ to be initialized anyhow? Something seems to be wrong here, perhaps we need separate types for "in" and "out" interfaces? Because we know that those that we get from native code always have these values. Or maybe we should generate classes implementing these interfaces that can be created from TS side using the appropriate ctors?

Comment (sergioferreiradt) [reply Source/Modules/javascript.cxx]: Your arguments make sense. Let me think a bit more about it.

Comment (sergioferreiradt) [reply Source/Modules/javascript.cxx]: After doing way too many tests and created some examples and explanation, I just realized that I was wanting to do a stupid thing: creating opscore objects from a literal initialization, such as:

let detailsCountry : DetailsCountry = {
  id: "PT",
  name: "Portugal",
  currency: "PTE",
};

instead of instantiating first by using the proper contructor existing in generated c++ wrapper: 

let detailsCountry : DetailsCountry = new datastore.DetailsCountry();
detailsCountry.id = 'PT';
detailsCountry.name = 'Portugal';
detailsCountry.currency = 'PTE';

Literal initialization of objects is pretty common in TypeScript/JavaScript, but now (hope not change again my opinion) I think it should be avoided because we are not creating a propper object that can be used against opscore classes.

So, in conclusion, I will change the function emition to not having optional functions that of course is a anacronic thing.

File: Source/Modules/javascript.cxx [hunk 5]
```diff
@@ -2843,7 +2838,15 @@ int TypeScriptTypes::enumvalueDeclaration(Node *n)
...
  * Handler executed in tree navigation when it encounters a class
- * member function
+ * member function.
+ *
+ * For TypeScript emits the prototype for the function wth returning type,
+ * parameters and corresponding parameters.
+ *
+ * As an example, for a C++ public function declared as:
+ *   int sum(int a,int b) { return a+b;}
+ * TypeScript emited will be:
```
Comment (vadz) [inline Source/Modules/javascript.cxx:2848]:
```suggestion
 * the prototype emitted for TypeScript will be:
```

File: Examples/test-suite/typescript_methods.i [hunk 6]
```diff
@@ -0,0 +1,18 @@
...
+
+class A {
+  std::string strValue;
+};
+
+class C {
+  public:
+    int intValue;
+    float floatValue;
+    int sum(int a,int b) { return a+b;}
```
Comment (vadz) [inline Examples/test-suite/typescript_methods.i:13]: Tiniest possible nitpick
```suggestion
    int sum(int a,int b) { return a+b; }
```

**Review by vadz — APPROVED** [review]
Looks nice and simple to me, Thanks!
